### ISSUE TITLE
"Corrupted MAC on input" errors using SFTP umac-64@openssh.com digest

### ISSUE THREAD

**Issue by cedef** [proftpd/proftpd#1111]
### What I Did

Using:
- latest proftpd version (build from `master`) with sftp module, with default Digests (MAC) algorithms
- /usr/bin/sftp client from openssh-client (`7.6p1-4ubuntu0.3`)

I'm connecting using umac-64 (which seems to be the default for both server & client with both no explicit configuration) or umac128 MAC algorithm:
```
sftp -v -F /dev/null -o MACs=umac-64@openssh.com -P 22 foobar@W.X.Y.Z
...[cut some irrelevant info]
debug1: Reading configuration data /dev/null
debug1: Connecting to W.X.Y.Z [W.X.Y.Z] port 22.
debug1: Connection established.
debug1: Authentications that can continue: password
debug1: Next authentication method: password
foobar@W.X.Y.Z's password:
debug1: Authentication succeeded (password).
Authenticated to W.X.Y.Z ([W.X.Y.Z]:22).
debug1: channel 0: new [client-session]
debug1: Entering interactive session.
Connected to W.X.Y.Z
sftp> ls data
data/anydir01  data/anydir02  data/anydir03  data/anydir04  data/anydir05  data/anydir06  data/anydir07
```
(So far, everything went fine).

Now, if I run `mkdir /home/foobar/data/anydir08 && chown foobar /home/foobar/data/anydir08` on the server, when listing `data` directory content I get:
```
Connected to W.X.Y.Z.
sftp> ls
data  
sftp> ls data
Corrupted MAC on input.
ssh_dispatch_run_fatal: Connection to W.X.Y.Z port 22: message authentication code incorrect
Connection closed
```

And on the server side's proftpd sftp.log:
```
2020-09-15 15:03:34,077 mod_sftp/1.0.1[19761]: sent server version 'SSH-2.0-mod_sftp'
2020-09-15 15:03:34,077 mod_sftp/1.0.1[19761]: received client version 'SSH-2.0-OpenSSH_7.6p1 Ubuntu-4ubuntu0.3'
2020-09-15 15:03:34,077 mod_sftp/1.0.1[19761]: handling connection from SSH2 client 'OpenSSH_7.6p1'
2020-09-15 15:03:34,116 mod_sftp/1.0.1[19761]:  + Session key exchange: curve25519-sha256
2020-09-15 15:03:34,116 mod_sftp/1.0.1[19761]:  + Session server hostkey: rsa-sha2-512
2020-09-15 15:03:34,116 mod_sftp/1.0.1[19761]:  + Session client-to-server encryption: aes128-ctr
2020-09-15 15:03:34,116 mod_sftp/1.0.1[19761]:  + Session server-to-client encryption: aes128-ctr
2020-09-15 15:03:34,116 mod_sftp/1.0.1[19761]:  + Session client-to-server MAC: umac-64@openssh.com
2020-09-15 15:03:34,116 mod_sftp/1.0.1[19761]:  + Session server-to-client MAC: umac-64@openssh.com
2020-09-15 15:03:34,116 mod_sftp/1.0.1[19761]:  + Session client-to-server compression: none
2020-09-15 15:03:34,116 mod_sftp/1.0.1[19761]:  + Session server-to-client compression: none
2020-09-15 15:03:34,257 mod_sftp/1.0.1[19761]: sending acceptable userauth methods: password
2020-09-15 15:03:36,541 mod_sftp/1.0.1[19761]: sending userauth success
2020-09-15 15:03:36,541 mod_sftp/1.0.1[19761]: user 'foobar' authenticated via 'password' method
2020-09-15 15:03:36,608 mod_sftp/1.0.1[19761]: 'subsystem' channel request for 'sftp' subsystem
2020-09-15 15:03:36,609 mod_sftp/1.0.1[19761]: using SFTP protocol version 3 for this session (channel ID 0)
2020-09-15 15:03:40,207 mod_sftp/1.0.1[19761]: disconnecting client (received EOF)
2020-09-15 15:03:40,210 mod_sftp/1.0.1[19761]: aborting 1 unclosed file handle
```

### What I Expected/Wanted

I expected sftp client `ls data/` command to list directory content, no matter the number of directory entries.
When switching to `hmac-sha2-512` MAC algorithm on the client side, I'm able to list the dir as expected:
```
sftp -v -F /dev/null -o MACs=hmac-sha2-512 -P 22 foobar@W.X.Y.Z
```

```
2020-09-15 15:11:35,097 mod_sftp/1.0.1[20201]: sent server version 'SSH-2.0-mod_sftp'
2020-09-15 15:11:35,097 mod_sftp/1.0.1[20201]: received client version 'SSH-2.0-OpenSSH_7.6p1 Ubuntu-4ubuntu0.3'
2020-09-15 15:11:35,097 mod_sftp/1.0.1[20201]: handling connection from SSH2 client 'OpenSSH_7.6p1'
2020-09-15 15:11:35,133 mod_sftp/1.0.1[20201]:  + Session key exchange: curve25519-sha256
2020-09-15 15:11:35,133 mod_sftp/1.0.1[20201]:  + Session server hostkey: rsa-sha2-512
2020-09-15 15:11:35,133 mod_sftp/1.0.1[20201]:  + Session client-to-server encryption: aes128-ctr
2020-09-15 15:11:35,133 mod_sftp/1.0.1[20201]:  + Session server-to-client encryption: aes128-ctr
2020-09-15 15:11:35,133 mod_sftp/1.0.1[20201]:  + Session client-to-server MAC: hmac-sha2-512
2020-09-15 15:11:35,133 mod_sftp/1.0.1[20201]:  + Session server-to-client MAC: hmac-sha2-512
2020-09-15 15:11:35,133 mod_sftp/1.0.1[20201]:  + Session client-to-server compression: none
2020-09-15 15:11:35,133 mod_sftp/1.0.1[20201]:  + Session server-to-client compression: none
2020-09-15 15:11:35,274 mod_sftp/1.0.1[20201]: sending acceptable userauth methods: password
2020-09-15 15:11:37,452 mod_sftp/1.0.1[20201]: sending userauth success
2020-09-15 15:11:37,452 mod_sftp/1.0.1[20201]: user 'foobar' authenticated via 'password' method
2020-09-15 15:11:37,520 mod_sftp/1.0.1[20201]: 'subsystem' channel request for 'sftp' subsystem
2020-09-15 15:11:37,520 mod_sftp/1.0.1[20201]: using SFTP protocol version 3 for this session (channel ID 0)
````

### ProFTPD Version and Configuration

```bash
proftpd -V
Compile-time Settings:
  Version: 1.3.8rc1 (git)
  Platform: LINUX [Linux 5.3.0-1011-gcp x86_64]
  Built: Tue Sep 15 2020 13:18:52 UTC
  Built With:
    configure  '--build=x86_64-linux-gnu' '--includedir=${prefix}/include' '--mandir=${prefix}/share/man' '--infodir=${prefix}/share/info' '--sysconfdir=/etc' '--localstatedir=/var' '--disable-silent-rules' '--libdir=${prefix}/lib/x86_64-linux-gnu' '--libexecdir=${prefix}/lib/x86_64-linux-gnu' '--disable-maintainer-mode' '--disable-dependency-tracking' '--prefix=/usr' '--with-includes=/usr/include/postgresql:/usr/include/mysql' '--mandir=/usr/share/man' '--sysconfdir=/etc/proftpd' '--localstatedir=/run' '--libexecdir=/usr/lib/proftpd' '--enable-sendfile' '--enable-facl' '--enable-dso' '--enable-autoshadow' '--enable-ctrls' '--enable-ipv6' '--enable-nls' '--enable-memcache' '--with-lastlog=/var/log/lastlog' '--enable-pcre' '--disable-strip' '--enable-redis' '--build' 'x86_64-linux-gnu' '--with-shared=mod_unique_id:mod_site_misc:mod_load:mod_ban:mod_quotatab:mod_sql:mod_sql_mysql:mod_sql_postgres:mod_sql_sqlite:mod_sql_odbc:mod_dynmasq:mod_quotatab_sql:mod_ldap:mod_quotatab_ldap:mod_ratio:mod_tls:mod_rewrite:mod_radius:mod_wrap:mod_wrap2:mod_wrap2_file:mod_wrap2_sql:mod_quotatab_file:mod_quotatab_radius:mod_facl:mod_ctrls_admin:mod_copy:mod_deflate:mod_ifversion:mod_sftp:mod_sftp_pam:mod_sftp_sql:mod_shaper:mod_sql_passwd:mod_ifsession:mod_auth_otp:mod_tls_redis:mod_wrap2_redis:mod_redis:mod_memcache:mod_tls_memcache:mod_readme:mod_snmp' 'build_alias=x86_64-linux-gnu'

  CFLAGS: -g2 -O2 -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration
  LDFLAGS: -L$(top_srcdir)/lib -L$(top_builddir)/lib  -rdynamic  -L/usr/lib/x86_64-linux-gnu -L/usr/lib/x86_64-linux-gnu
  LIBS:  -lpcreposix -lpcre -lssl -lcrypto -lsupp -lnsl -lresolv -lresolv -lcrypt -ldl -lmemcachedutil -lmemcached  -pthread

  Files:
    Configuration File:
      /etc/proftpd/proftpd.conf
    Pid File:
      /run/proftpd.pid
    Scoreboard File:
      /run/proftpd.scoreboard
    Header Directory:
      /usr/include/proftpd
    Shared Module Directory:
      /usr/lib/proftpd

  Info:
    + Max supported UID: 4294967295
    + Max supported GID: 4294967295

  Features:
    + Autoshadow support
    + Controls support
    - curses support
    - Developer support
    + DSO support
    + IPv6 support
    + Largefile support
    + Lastlog support
    + Memcache support
    - ncurses support
    + NLS support
    + OpenSSL support (OpenSSL 1.1.1c  28 May 2019)
    + PCRE support
    + POSIX ACL support
    - Redis support
    + Sendfile support
    + Shadow file support
    - Sodium support
    + Trace support
    + xattr support

  Tunable Options:
    PR_TUNABLE_BUFFER_SIZE = 1024
    PR_TUNABLE_DEFAULT_RCVBUFSZ = 8192
    PR_TUNABLE_DEFAULT_SNDBUFSZ = 8192
    PR_TUNABLE_ENV_MAX = 2048
    PR_TUNABLE_GLOBBING_MAX_MATCHES = 100000
    PR_TUNABLE_GLOBBING_MAX_RECURSION = 8
    PR_TUNABLE_HASH_TABLE_SIZE = 40
    PR_TUNABLE_LOGIN_MAX = 256
    PR_TUNABLE_NEW_POOL_SIZE = 512
    PR_TUNABLE_PATH_MAX = 4096
    PR_TUNABLE_SCOREBOARD_BUFFER_SIZE = 80
    PR_TUNABLE_SCOREBOARD_SCRUB_TIMER = 30
    PR_TUNABLE_SELECT_TIMEOUT = 30
    PR_TUNABLE_TIMEOUTIDENT = 10
    PR_TUNABLE_TIMEOUTIDLE = 600
    PR_TUNABLE_TIMEOUTLINGER = 10
    PR_TUNABLE_TIMEOUTLOGIN = 300
    PR_TUNABLE_TIMEOUTNOXFER = 300
    PR_TUNABLE_TIMEOUTSTALLED = 3600
    PR_TUNABLE_XFER_SCOREBOARD_UPDATES = 10
```

My proftpd.conf file:
```conf
ServerType				standalone
DefaultServer			on
Port				21

MaxInstances			30

User				proftpd
Group				nogroup

Umask				002  002

TransferLog /var/log/proftpd/xferlog
SystemLog   /var/log/proftpd/proftpd.log
TraceLog /var/log/proftpd/trace.log
Trace DEFAULT:10
LoadModule mod_sftp.c
LoadModule mod_sftp_pam.c

SFTPEngine on
Port 22

DefaultRoot ~
RequireValidShell off
SFTPPAMEngine On

SFTPLog /var/log/proftpd/sftp.log
SFTPHostKey /etc/proftpd/sftp_host_rsa_key
SFTPAuthMethods password
```

**Comment by Castaglia** [comment]
I'm attempting to reproduce this locally, using a Docker `ubuntu:18.04` image.

Does the directory creation, `chown` need to happen outside of the SFTP connection, for this to trigger?  Does it happen only after a certain amount of time/data?

**Comment by cedef** [comment]
> However, the "curve25519-sha256" key exchange algorithm is only available if libsodium is installed/available/used -- but your proftpd -V output shows no libsodium support?

I've fixed that by installing libsodium. Here the new [proftpd -V](https://github.com/proftpd/proftpd/files/5237771/proftpd-version.txt) output. I still hit the bug :cry: 

> Does the directory creation, chown need to happen outside of the SFTP connection, for this to trigger? 

Actually, there is no need of directory creation: I've talked about this detail because I _think_ that the problem might be "memory" related (maybe similar to what have been found in #1060 ?) : with 7 directories (anydir01 to 07) I'm able to run the sftp command `ls data`. With 8 directories, I'm kicked out with `Corrupted MAC on input.` no matter if I create or not this directory while connected on the SFTP or not. It _looks like_ it is related to the result size of the `ls` command...

Same for the `chown` command: it's just to avoid rights/authorization stuff from the bug reproducibility : for what I've seen so far, it's not linked to any permission denied problem or so. I've tried to reproduce the problem with minimum configuration, through simple example.

> Does it happen only after a certain amount of time/data?

There is no need to wait any amount of time. For what I know, just listing the directory content is enough to reproduce the bug. Even this command triggers the bug:
```
echo ls data | sftp -v -F /dev/null -o MACs=umac-64@openssh.com -P 22 foobar@W.X.Y.Z
```

Please let me know if you need more details.
Regards,

**Comment by Castaglia** [comment]
I'm still not having any luck reproducing this locally, using Docker.  I've created 12 directories for my test user, with longer names -- still no corruption.   Hrm.

One difference I note is that, with the `ubuntu:18.04` image I'm using, I have a slightly different OpenSSL version:
```
# /usr/local/sbin/proftpd -V
...
    + OpenSSL support (OpenSSL 1.1.1  11 Sep 2018)
```
I wonder if that makes a difference?  I'll see where I can find an OpenSSL-1.1.1c package for Ubuntu 18.04.

**Comment by Castaglia** [comment]
There was a memory-corruption bug in `mod_sftp`, [Bug #4405](http://bugs.proftpd.org/show_bug.cgi?id=4405) -- but that was fixed in August; I'm assuming you compiled ProFTPD from source _after_ that date.

**Comment by locojohn** [comment]
For me it is the same situation and this bug is always reproducible when trying to list contents of a directory.  I compiled Proftpd 1.3.7a from source (Gentoo Linux user here).

**Comment by severach** [comment]
Same problem here in Arch Linux. `ls` works, `get` or `put` disconnects with error. Switch to `MACs=hmac-sha2-256` and everything works.

Tried Linux FileZilla, works fine. Tried Linux PuTTY psftp, works fine. Tried `proftpd-git-1.3.7.r159.g0d997e50e-1`, crashes. Tried turning off all the modules, crashes. Try configure --disable a bunch of stuff, crashes.

It's a gcc 10 problem. Compile with gcc 9 or clang fixes the problem. `sshfs` is working perfectly. Time to add compiler version to `proftpd -V`.

````
# gcc-10
[chris@ga ~]$ sftp -F /dev/null -o MACs=hmac-sha2-256 -P 2222 chris@128.0.0.6
chris@128.0.0.6's password: 
Connected to 128.0.0.6.
sftp> ls
lpstat-m.ga  
sftp> get lpstat-m.ga
Fetching /lpstat-m.ga to lpstat-m.ga
/lpstat-m.ga                                  100%   48KB   6.0MB/s   00:00    
sftp> put lpstat-m.ga
Uploading lpstat-m.ga to /lpstat-m.ga
/lpstat-m.ga                                  100%   48KB   6.0MB/s   00:00    
sftp> quit

[chris@ga ~]$ sftp -F /dev/null -o MACs=umac-64@openssh.com -P 2222 chris@128.0.0.6
chris@128.0.0.6's password: 
Connected to 128.0.0.6.
sftp> ls
lpstat-m.ga  
sftp> get lpstat-m.ga
Fetching /lpstat-m.ga to lpstat-m.ga
Corrupted MAC on input.
ssh_dispatch_run_fatal: Connection to 128.0.0.6 port 2222: message authentication code incorrect
Connection closed
````

SFTPLog contain this. The garbage text after "MAC using" changes periodically.
````
mod_sftp/1.0.1[2850559]: MAC from client differs from expected MAC using +#K^E.f^F
mod_sftp/1.0.1[2850559]: unable to verify MAC on packet from socket 0
mod_sftp/1.0.1[2850559]: disconnecting 128.0.0.202 (Application disconnected)
````

````
# gcc-9 or clang
[chris@ga ~]$ sftp -F /dev/null -o MACs=umac-64@openssh.com -P 2222 chris@128.0.0.6
chris@128.0.0.6's password: 
Connected to 128.0.0.6.
sftp> ls
lpstat-m.ga  
sftp> get lpstat-m.ga
Fetching /lpstat-m.ga to lpstat-m.ga
/lpstat-m.ga                                  100%   48KB   6.0MB/s   00:00    
sftp> put lpstat-m.ga
Uploading lpstat-m.ga to /lpstat-m.ga
lpstat-m.ga                                   100%   48KB   6.2MB/s   00:00    
sftp> quit
````

````
# pacman -Q proftpd gcc gcc9 openssl binutils clang libsodium openssh
proftpd 2:1.3.7a-2
gcc 10.2.0-6
gcc9 9.3.0-5
openssl 1.1.1.j-1
binutils 2.36.1-2
clang 11.1.0-1
libsodium 1.0.18-2
openssh 8.4p1-2
````

````
# proftpd -V
Compile-time Settings:
  Version: 1.3.7a (maint)
  Platform: LINUX [Linux 4.19.176-1-lts419 x86_64]
  Built: Wed Feb 24 2021 02:57:04 UTC
  Built With:
    configure  '--prefix=/usr' '--sbindir=/usr/bin' '--libexecdir=/usr/lib' '--disable-pam' '--with-modules=mod_digest:mod_sftp:mod_quotatab:mod_quotatab_sql:mod_quotatab_file:mod_tls:mod_tls_shmcache:mod_ldap:mod_sql:mod_sql_mysql:mod_sql_postgres:mod_facl:mod_dynmasq' '--sysconfdir=/etc' '--localstatedir=/run/proftpd' '--enable-ctrls' '--enable-ipv6' '--with-includes=/usr/include/mysql:/usr/include/postgresql' '--with-libraries=/usr/lib/mysql:/usr/lib/postgresql' '--enable-nls' '--enable-redis' '--enable-facl' '--with-systemd' 'CC=gcc-9' 'CXX=g++-9' 'CFLAGS=-march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fno-plt' 'LDFLAGS=-Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now' 'CPPFLAGS=-D_FORTIFY_SOURCE=2' 'CXXFLAGS=-march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fno-plt'

  CFLAGS: -g2 -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fno-plt -Wall -fno-omit-frame-pointer
  LDFLAGS: -L$(top_srcdir)/lib -L$(top_builddir)/lib -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -rdynamic -L/usr/lib/postgresql -L/usr/lib/mysql -L/usr/lib/ -L/usr/lib
  LIBS: -lacl  -lssl -lcrypto -lsodium -lcap  -lpq  -lm -lmysqlclient -lz  -lldap -llber  -lssl  -lpam  -lcrypto -lsupp -lattr -lcrypt -ldl -lhiredis  -pthread

  Files:
    Configuration File:
      /etc/proftpd.conf
    Pid File:
      /run/proftpd/proftpd.pid
    Scoreboard File:
      /run/proftpd/proftpd.scoreboard

  Info:
    + Max supported UID: 4294967295
    + Max supported GID: 4294967295

  Features:
    - Autoshadow support
    + Controls support
    + curses support
    - Developer support
    - DSO support
    + IPv6 support
    + Largefile support
    - Lastlog support
    - Memcache support
    + ncursesw support
    + NLS support
    + OpenSSL support (OpenSSL 1.1.1j  16 Feb 2021)
    - PCRE support
    + POSIX ACL support
    + Redis support
    + Sendfile support
    + Shadow file support
    + Sodium support
    + Trace support
    + xattr support

  Tunable Options:
    PR_TUNABLE_BUFFER_SIZE = 1024
    PR_TUNABLE_DEFAULT_RCVBUFSZ = 8192
    PR_TUNABLE_DEFAULT_SNDBUFSZ = 8192
    PR_TUNABLE_ENV_MAX = 2048
    PR_TUNABLE_GLOBBING_MAX_MATCHES = 100000
    PR_TUNABLE_GLOBBING_MAX_RECURSION = 8
    PR_TUNABLE_HASH_TABLE_SIZE = 40
    PR_TUNABLE_LOGIN_MAX = 256
    PR_TUNABLE_NEW_POOL_SIZE = 512
    PR_TUNABLE_PATH_MAX = 4096
    PR_TUNABLE_SCOREBOARD_BUFFER_SIZE = 80
    PR_TUNABLE_SCOREBOARD_SCRUB_TIMER = 30
    PR_TUNABLE_SELECT_TIMEOUT = 30
    PR_TUNABLE_TIMEOUTIDENT = 10
    PR_TUNABLE_TIMEOUTIDLE = 600
    PR_TUNABLE_TIMEOUTLINGER = 10
    PR_TUNABLE_TIMEOUTLOGIN = 300
    PR_TUNABLE_TIMEOUTNOXFER = 300
    PR_TUNABLE_TIMEOUTSTALLED = 3600
    PR_TUNABLE_XFER_SCOREBOARD_UPDATES = 10
````

**Comment by Castaglia** [comment]
@severach thanks for digging into this!  Compiler bugs are indeed tricky.  I'll see if I can reproduce this using an Arch Linux Docker image; I've had trouble reproducing this locally -- but knowing that it's related to the compiler version will help.  I'm not sure there'll be a workaround, but I can try.

**Comment by Castaglia** [comment]
@severach I don't normally use Archlinux (I develop on a Mac), so I usually use Docker for working with Linux platforms.  Do you happen to have a Dockerfile, for your Archlinux environment, that I might use for reproducing this behavior locally?

**Comment by locojohn** [comment]
no dockerfile.  Here is gcc configuration:

# gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-pc-linux-gnu/10.2.0/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /var/tmp/portage/sys-devel/gcc-10.2.0-r5/work/gcc-10.2.0/configure --host=x86_64-pc-linux-gnu --build=x86_64-pc-linux-gnu --prefix=/usr --bindir=/usr/x86_64-pc-linux-gnu/gcc-bin/10.2.0 --includedir=/usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0/include --datadir=/usr/share/gcc-data/x86_64-pc-linux-gnu/10.2.0 --mandir=/usr/share/gcc-data/x86_64-pc-linux-gnu/10.2.0/man --infodir=/usr/share/gcc-data/x86_64-pc-linux-gnu/10.2.0/info --with-gxx-include-dir=/usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0/include/g++-v10 --with-python-dir=/share/gcc-data/x86_64-pc-linux-gnu/10.2.0/python --enable-languages=c,c++ --enable-obsolete --enable-secureplt --disable-werror --with-system-zlib --enable-nls --without-included-gettext --enable-checking=release --with-bugurl=https://bugs.gentoo.org/ --with-pkgversion='Gentoo 10.2.0-r5 p6' --disable-esp --enable-libstdcxx-time --enable-shared --enable-threads=posix --enable-__cxa_atexit --enable-clocale=gnu --disable-multilib --with-multilib-list=m64 --disable-fixed-point --enable-targets=all --enable-libgomp --disable-libssp --disable-libada --disable-systemtap --enable-vtable-verify --without-zstd --enable-lto --without-isl --enable-default-pie --enable-default-ssp
Thread model: posix
Supported LTO compression algorithms: zlib
gcc version 10.2.0 (Gentoo 10.2.0-r5 p6)

# openssl version
OpenSSL 1.1.1i  8 Dec 2020

**Comment by Castaglia** [comment]
No worries.

I think I managed to get a local setup that reproduces this bug, using an Ubuntu 18.04 Docker image and a `gcc-10` package from a PPA.  Progress, at least!

**Comment by Castaglia** [comment]
The current `mod_sftp/umac.c` file is largely borrowed from the OpenSSH implementation.  So I started to wonder why OpenSSH does not have these bug reports, even with gcc-10?

In looking at the upstream changes to that `umac.c` source code, and -- more importantly -- the compiler flags that OpenSSH uses (_vs_ ProFTPD), I managed to find the one compiler flag that "fixes" this bug (in addition to the upstream `umac.c` changes).

@severach @locojohn could you thus try out [this PR](https://github.com/proftpd/proftpd/pull/1181), verify that it fixes things for your environments?

**Comment by Castaglia** [comment]
This should now be fixed in master; the fix was backported to the 1.3.7 branch as well.  Thanks!
